### PR TITLE
Shut down mapgen threads after kicking players and before other shutdown tasks

### LINKED ISSUE #6680: mod_storage unreliable when used in register_on_generated()
I was experimenting with capturing data from mapgen objects that are not available outside the mapgen thread, and storing this data in mod_storage.

The good part is that mod_storage functions and is entirely available inside the mapgen thread, and data can indeed get written to the mod_storage.

The bad part is that the storage is unreliable, as it seems that the mod_storage object goes away when the server exits, *before* all the mapgen threads are able to shut down and write their data to the mod_storage.

The result is that only *some* of the data successfully gets written into the mod_storage.

Due to the unreliabilty of mod_storage, it is effectively unusable for many tasks, if not all. Some map data is only accessible during mapgen threads, and losing it randomly would make the data lost forever.

To fix, the mapgen threads need to be shut down *before* the main server thread, not before.

### PR DESCRIPTION
**Pull request by raymoo**
Fixes #6680.

Also stops mapgen threads before calling `register_on_shutdown` callbacks, in case something modified by mapgen threads are saved in the shutdown callback.

I only tested that no mapgen callbacks run after the shutdown callbacks. I did not test the functionality of ModStorage threads in mapgen callbacks, but if it is available in shutdown callbacks, and shutdown callbacks happen after the mapgen threads finish, then it should be available in mapgen too. Here is the code I used to test:
```
local bob = false

minetest.register_on_shutdown(function()
        bob = true
end)

minetest.register_on_generated(function()
        if bob then
		print("MEMES!!!!")
        else
                assert(false)
        end
end)
```
To test, play a map a few times with the mod, running Minetest from the command line, and make sure to go somewhere that the map will be generated. If you see the message at least once, the fix didn't work.

### CODE REVIEW

**Comment by sofar** [comment]
BTW, this may make shutdown take significantly longer (at least until players are kicked).

**Comment by paramat** [comment]
Longer seems acceptable to shutdown properly.

**Comment by raymoo** [comment]
@sofar Would that be solved by just shutting down the mapgen threads in between kicking players and the other stuff?

**Comment by sofar** [comment]
No, I think it's fine, it's just that it may surprise players if a significant amount of mapgen emerges are happening.

**Comment by rubenwardy** [comment]
I'd kick the players, then shut down mapgen

**Comment by raymoo** [comment]
@rubenwardy It wouldn't make sense to me if a shutdown callback relied on the players being still online, because what if the players had all just logged off normally? I will try moving the mapgen finishing below the kicking.

**Comment by raymoo** [comment]
Addressed @rubenwardy's comment.

File: src/server.cpp [hunk 1]
```diff
@@ -333,6 +330,22 @@ Server::~Server()
 		}
 		m_env->kickAllPlayers(SERVER_ACCESSDENIED_SHUTDOWN,
 			kick_msg, reconnect);
+	}
+
+	// Stop all emerge threads before deleting players that may have requested
+	// blocks to be emerged.
```
Comment (SmallJoker) [inline src/server.cpp]: No longer valid.

File: src/server.cpp [hunk 2]
```diff
@@ -333,6 +330,22 @@ Server::~Server()
...
+	// server-controlled resources (like ModStorages). Also do them before
+	// shutdown callbacks since they may modify state that is finalized in a
+	// callback.
+	m_emerge->stopThreads();
+
+	{
+		MutexAutoLock envlock(m_env_mutex);
+
+		// Execute script shutdown hooks
+		infostream<<"Executing shutdown hooks"<<std::endl;
```
Comment (SmallJoker) [inline src/server.cpp]: Needs spaces around operators: ` << `

File: src/server.cpp (hunk 1 above)
Comment (raymoo) [reply src/server.cpp]: Then it wasn't valid before either, since I moved it up. I don't know when it was last valid (will fix though)

**Comment by raymoo** [comment]
Responded to @SmallJoker's comments

**Comment by rubenwardy** [comment]
Confirmed, the comment was from before the players were kicked: https://github.com/minetest/minetest/commit/9b978db0c2578b06c5669096e325c1ce70864edc#diff-ad60d65b34e16a3319296bb5d683acd6L920

**Comment by raymoo** [comment]
@rubenwardy Thanks for checking it out.

**Pull request merged**